src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs

```diff
@@ -32,68 +33,93 @@ public partial class EventSourceGenerator : ISourceGenerator
         //         }
         //     }
 
-        public void Initialize(GeneratorInitializationContext context)
-            => context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValuesProvider<EventSourceClass> eventSourceClasses =
+                context.SyntaxProvider
+                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), (x, _) => GetSemanticTargetForGeneration(x))
+                .Where(x => x is not null)
+                .Collect()
+                .Combine(context.CompilationProvider)
+                // Use a custom comparer that ignores the compilation.
+                .WithComparer(ClassDeclarationsEqualityComparer.Instance)
+                .SelectMany((x, cancellationToken) => Parser.GetEventSourceClasses(x.Left, x.Right, cancellationToken));
+
+            context.RegisterSourceOutput(eventSourceClasses, static (spc, x) => Emitter.Emit(spc, x));
+        }
 
-        public void Execute(GeneratorExecutionContext context)
+        private static bool IsSyntaxTargetForGeneration(SyntaxNode node) =>
+            node is ClassDeclarationSyntax x && x.AttributeLists.Count > 0;
+
+        private static ClassDeclarationSyntax? GetSemanticTargetForGeneration(GeneratorSyntaxContext context)
```

Comment:
This doesn't appear to be actually doing any semantic work? It's just checking the syntax. In which case you should move the logic into the syntactic check.

Syntax checks are only re-performed on files that are modified, whereas semantic checks have to run every time any edit is made (as you might be looking something up in another file). Moving this work to the syntactic check will give you significantly improved perf in the IDE on bigger projects.

Reply:
So I should move `GetSemanticTargetForGeneration` into `IsSyntaxTargetForGeneration`? [I read that](https://andrewlock.net/creating-a-source-generator-part-1-creating-an-incremental-source-generator/) `CreateSyntaxProvider`'s predicate must be very fast, but it looks like that the litmus test is whether we use the context's `SemanticModel`. Thanks, I will fix it.

Reply:
(or even better I will actually use the semantic model and correctly detect the attribute)